dataplane/ebpf/src/ingress/tcp.rs

```diff
@@ -116,21 +126,106 @@ pub fn handle_tcp_ingress(ctx: TcContext) -> Result<i32, i64> {
         return Ok(TC_ACT_OK);
     }
 
-    // Calculate l3 cksum
-    // TODO(astoycos) use l3_cksum_replace instead
-    unsafe { (*ip_hdr).check = 0 };
-    let full_cksum = unsafe {
-        bpf_csum_diff(
-            mem::MaybeUninit::zeroed().assume_init(),
+    let mut lb_mapping = LoadBalancerMapping {
+        backend,
+        backend_key,
+        tcp_state,
+    };
+
+    let tcp_hdr_ref = unsafe { tcp_hdr.as_ref().ok_or(TC_ACT_OK)? };
+
+    // If the packet has the RST flag set, it means the connection is being terminated, so remove it
+    // from our map.
+    if tcp_hdr_ref.rst() == 1 {
+        unsafe {
+            LB_CONNECTIONS.remove(&client_key)?;
+        }
+    }
+
+    update_tcp_conns(tcp_hdr_ref, &client_key, &mut lb_mapping)?;
+
+    // inspired by https://github.com/torvalds/linux/blob/master/samples/bpf/tcbpf1_kern.c
+    // update dst_addr in the ip_hdr
+    // recalculate the checksums
+    unsafe {
+        let backend_ip = backend.daddr.to_be();
+
+        let ret = bpf_l4_csum_replace(
+            ctx.skb.skb,
+            TCP_CSUM_OFF,
+            original_daddr as u64,
+            backend_ip as u64,
+            IS_PSEUDO | (mem::size_of_val(&backend_ip) as u64),
```

Comment:
What is the `IS_PSEUDO` const doing here?

Reply:
According to my understanding, `IS_PSEUDO | (mem::size_of_val(&backend_ip) as u64)` as the last argument of `bpf_l4_csum_replace()`, it can simultaneously set the IS_PSEUDO flag and specify the size of the field to be updated, allowing the function to correctly recalculate the checksum of the transport layer and consider the impact of the pseudo header.

see more:

1. result of `IS_PSEUDO | (mem::size_of_val(&backend_ip) as u64)`: `00010000 | 0100` -> `00010100`

2. in [`bpf_l4_csum_replace()`](https://elixir.bootlin.com/linux/v5.18.19/source/net/core/filter.c#L1935) 

	```c
		switch (flags & BPF_F_HDR_FIELD_MASK) {
		case 0:
			if (unlikely(from != 0))
				return -EINVAL;
	
			inet_proto_csum_replace_by_diff(ptr, skb, to, is_pseudo);
			break;
		case 2:
			inet_proto_csum_replace2(ptr, skb, from, to, is_pseudo);
			break;
		case 4:
			inet_proto_csum_replace4(ptr, skb, from, to, is_pseudo);
			break;
	```
	here `BPF_F_HDR_FIELD_MASK` is
    ```c
		enum {
		BPF_F_HDR_FIELD_MASK		= 0xfULL,
		};
	```

	`flags & BPF_F_HDR_FIELD_MASK`: `00010100 & 00001111 ` -> `00000100`, and would chosse `case 4` to update l4 checksum.